### PR TITLE
 authn-JWT support

### PR DESCRIPTION
**Pull request by quincycheng**
### Desired Outcome

Enable summon-conjur to authn Conjur with authn-jwt

### Implemented Changes

*Describe how the desired outcome above has been achieved with this PR. In
particular, consider:*

Introducing 2 new env variables `CONJUR_AUTHN_JWT_SERVICE_ID` &  `JWT_TOKEN_PATH`

### Connected Issue/Story

It's new.

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [X] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [X] The changes in this PR do not require tests

#### Documentation

- [X] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

**Comment by infamousjoeg** [comment]
I feel like this should be merged into a branch other than `main` to allow for additional testing prior to merge into `main` and official release.  Thoughts, @cyberark/community-and-integrations-team ?

File: conjurapi/client.go [hunk 1]
```diff
@@ -114,6 +114,61 @@ func NewClientFromEnvironment(config Config) (*Client, error) {
...
+	if authnTokenFile != "" {
+		return NewClientFromToken(config, authnToken)
+	}
+
+	authnJwtServiceID := os.Getenv("CONJUR_AUTHN_JWT_SERVICE_ID")
+	if authnJwtServiceID != "" {
+
+		jwtTokenPath := os.Getenv("JWT_TOKEN_PATH")
+		if jwtTokenPath == "" {
+			jwtTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
```
Comment (tzheleznyak) [inline conjurapi/client.go:127]: Can this string moved to const in another place?

File: conjurapi/client.go [hunk 2]
```diff
@@ -114,6 +114,61 @@ func NewClientFromEnvironment(config Config) (*Client, error) {
...
+		}
+
+		jwtToken, err := ioutil.ReadFile(jwtTokenPath)
+		if err != nil {
+			return nil, err
+		}
+		jwtTokenString := "jwt=" + string(jwtToken)
```
Comment (tzheleznyak) [inline conjurapi/client.go:134]: Please use fmt.Sprintf to format the string here

File: conjurapi/client.go [hunk 3]
```diff
@@ -114,6 +114,61 @@ func NewClientFromEnvironment(config Config) (*Client, error) {
 		return NewClientFromTokenFile(config, authnTokenFile)
 	}
 
+	authnToken := os.Getenv("CONJUR_AUTHN_TOKEN")
```
Comment (tzheleznyak) [inline conjurapi/client.go:117]: Mayne this code can be broken to few functions of
* Read JWT Token File
* Build request
* Send Request

**Comment by tzheleznyak** [comment]
Fix conflict

**Comment by tzheleznyak** [comment]
There is also a flow of authn-k8s here or other authenticatios in this api ? If so where is the decision making?

File: conjurapi/client.go [hunk 4]
```diff
@@ -114,6 +114,61 @@ func NewClientFromEnvironment(config Config) (*Client, error) {
...
+			httpClient, err = newHTTPSClient(cert)
+			if err != nil {
+				return nil, err
+			}
+
+		} else {
+			httpClient = &http.Client{Timeout: time.Second * 10}
+		}
+
+		authnJwtUrl := makeRouterURL(config.ApplianceURL, "authn-jwt", authnJwtServiceID, config.Account, "authenticate").String()
```
[the remark this answers is not in the document]
Comment (QuincyChengAtWork) [reply conjurapi/client.go:151]: Not sure if I understand "cases of identity" here.   Can you explain more about that?

File: conjurapi/client.go (hunk 1 above)
Comment (QuincyChengAtWork) [reply conjurapi/client.go:127]: Sure.   I can't find any files for variables or constant declaration.   Any suggestions?

File: conjurapi/client.go (hunk 3 above)
Comment (QuincyChengAtWork) [reply conjurapi/client.go:117]: Please feel free to refactor it :)

**Comment by QuincyChengAtWork** [comment]
> There is also a flow of authn-k8s here or other authenticatios in this api ? If so where is the decision making?

Nope, nothing specific to authn-k8s in this PR.   I believe there's a decision making flow at https://github.com/cyberark/conjur-api-go/blob/9f0d30dfffa0c07ce09dface49e24bf9e5a4ce3f/conjurapi/client.go#L112-L125 for determining which info are used to authn to conjur

**Comment by infamousjoeg** [comment]
@jtuttle @tzheleznyak Happy to help contribute to get this pushed through. Please let me know if there are any pending changes required to approve this PR.

File: conjurapi/client.go (hunk 2 above)
Comment (quincycheng) [reply conjurapi/client.go]: Updated

**Comment by quincycheng** [comment]
> @quincycheng Can you squash this down into a single, descriptive commit?

Sure, and it's been squashed

**Comment by quincycheng** [comment]
@jtuttle @tzheleznyak Thanks for reviewing and providing advices.   
I believe I have made all the changes to PR.   
I am wondering if you can either clarify your suggestions or approve the changes - some clients are pending for this enhancement (which ultimately enable summon to support authn-jwt).   Thanks

File: conjurapi/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tzheleznyak) [reply conjurapi/client.go:151]: In JWT authentication the host can be provided in one of these ways:
Option 1:  In URL between the conjur account and the word authenticate.
Optino 2: In JWT token using token-app-property variable. I think here only option 2 is supported.

[the remark this answers is not in the document]
Comment (QuincyChengAtWork) [reply conjurapi/client.go:151]: According to [doc](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Operations/Services/cjr-authn-jwt.htm?tocpath=Integrations%7CJWT%20Authenticator%7C_____0#Definerelationshipinpolicy), using `token-app-property` (option 2) is recommended for production env.   
Let's follow the official recommendation :)

**Comment by infamousjoeg** [comment]
So.... close... to... merge....

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Added
+- Add `CONJUR_AUTHN_JWT_SERVICE_ID` & `JWT_TOKEN_PATH` environment variables as configuration to support authn-jwt 
+
```
Comment (rpothier) [inline CHANGELOG.md:10]: Can you add a link to this PR? [cyberark/conjur-api-go#124](https://github.com/cyberark/conjur-api-go/pull/124)

**Review by rpothier — COMMENTED** [review]
Looks good! just one small comment.

File: CHANGELOG.md (hunk 5 above)
Comment (quincycheng) [reply CHANGELOG.md]: Sure.  Link added

**Pull request merged**